commandsv3/src/test/java/org/wpilib/commands3/SchedulerTest.java

```diff
@@ -0,0 +1,1176 @@
+    m_scheduler.schedule(cmd);
+    m_scheduler.run();
+    m_scheduler.schedule(interrupter);
+    m_scheduler.run();
+
+    assertTrue(ran.get(), "onCancel should have run!");
+  }
+
+  @Test
+  void doesNotRunOnCancelWhenCompleting() {
+    var ran = new AtomicBoolean(false);
+
+    var mechanism = new Mechanism("The mechanism", m_scheduler);
+    var cmd = mechanism.run(Coroutine::yield).whenCanceled(() -> ran.set(true)).named("cmd");
+    m_scheduler.schedule(cmd);
+    m_scheduler.run();
+    m_scheduler.run();
+
+    assertFalse(m_scheduler.isScheduledOrRunning(cmd));
+    assertFalse(ran.get(), "onCancel ran when it shouldn't have!");
+  }
+
+  @Test
+  void runsOnCancelWhenCanceling() {
+    var ran = new AtomicBoolean(false);
+
+    var mechanism = new Mechanism("The mechanism", m_scheduler);
+    var cmd = mechanism.run(Coroutine::yield).whenCanceled(() -> ran.set(true)).named("cmd");
+    m_scheduler.schedule(cmd);
+    m_scheduler.run();
+    m_scheduler.cancel(cmd);
+
+    assertTrue(ran.get(), "onCancel should have run!");
+  }
+
+  @Test
+  void runsOnCancelWhenCancelingParent() {
+    var ran = new AtomicBoolean(false);
+
+    var mechanism = new Mechanism("The mechanism", m_scheduler);
+    var cmd = mechanism.run(Coroutine::yield).whenCanceled(() -> ran.set(true)).named("cmd");
+
+    var group = new Sequence("Seq", Collections.singletonList(cmd));
+    m_scheduler.schedule(group);
+    m_scheduler.run();
+    m_scheduler.cancel(group);
+
+    assertTrue(ran.get(), "onCancel should have run!");
+  }
+
+  @Test
+  void sideloadThrowingException() {
+    m_scheduler.sideload(
+        co -> {
+          throw new RuntimeException("Bang!");
+        });
+
+    // An exception raised in a sideload function should bubble up
+    assertEquals(
+        "Bang!", assertThrowsExactly(RuntimeException.class, m_scheduler::run).getMessage());
+  }
+
+  @Test
+  void periodicSideload() {
+    AtomicInteger count = new AtomicInteger(0);
+    m_scheduler.addPeriodic(count::incrementAndGet);
+    assertEquals(0, count.get());
+
+    m_scheduler.run();
+    assertEquals(1, count.get());
+
+    m_scheduler.run();
+    assertEquals(2, count.get());
+  }
+
+  @Test
+  void sideloadSchedulingCommand() {
+    var command = Command.noRequirements().executing(Coroutine::park).named("Command");
+    // one-shot sideload forks a command and immediately exits
+    m_scheduler.sideload(co -> co.fork(command));
+    m_scheduler.run();
+    assertTrue(
+        m_scheduler.isRunning(command), "command should have started and outlasted the sideload");
+  }
+
+  @Test
+  void childCommandEscapesViaSideload() {
+    var child = Command.noRequirements().executing(Coroutine::park).named("Child");
+    var parent =
+        Command.noRequirements()
+            .executing(
+                parentCoroutine -> {
+                  m_scheduler.sideload(sidelodCoroutine -> sidelodCoroutine.fork(child));
```

Comment:
Won't this schedule each iteration?

Additional tests that I thought of atm:
- await from sideload (I want to see that it doesn't block or do other weird stuff)
- cancel/interrupt a sideload-scheduled command (test hierarchy traversal logic)